### PR TITLE
Codechange: Replace FOR_ALL_XXX with range-based for loops

### PR DESCRIPTION
**Pull request by glx22**

### CODE REVIEW

File: src/core/pool_type.hpp [hunk 1]
```diff
@@ -284,6 +329,13 @@ struct Pool : PoolBase {
 		 * @note it's called only when !CleaningPool()
 		 */
 		static inline void PostDestructor(size_t index) { }
+
+		/**
+		 * Returns an iterable ensemble of all valid Titem
+		 * @param from index of the first Titem to consider
+		 * @return an iterable ensemble of all valid Titem
+		 */
+		static Pool::IterateWrapper<Titem> Iterate(size_t from = 0) { return Pool::IterateWrapper<Titem>(from); }
```
Comment (LordAro) [inline src/core/pool_type.hpp:340]: perhaps a specific const variant could be useful as well?

File: src/core/pool_type.hpp [hunk 2]
```diff
@@ -136,6 +137,47 @@ struct Pool : PoolBase {
...
+	 * Iterator to iterate all valid T of a pool
+	 * @tparam T Type of the class/struct that is going to be iterated
+	 */
+	template <class T>
+	struct PoolIterator {
+		typedef T value_type;
+		typedef T* pointer;
+		typedef T& reference;
+		typedef size_t difference_type;
+		typedef std::forward_iterator_tag iterator_category;
```
Comment (LordAro) [inline src/core/pool_type.hpp:150]: these all seem to be unused? Or are they necessary for the iterator "type" ?

**Review by LordAro — COMMENTED** [review]
:+1:

File: src/core/pool_type.hpp (hunk 2 above)
Comment (glx22) [reply src/core/pool_type.hpp:150]: They are necessary if we want to use the iterator in std algorithms like std::distance() or std::count_if().
But they are not needed for range-based for loops.

File: src/network/network.cpp [hunk 3]
```diff
@@ -99,8 +99,7 @@ extern void StateGameLoop();
  */
 bool HasClients()
 {
-	NetworkClientSocket *cs;
-	FOR_ALL_CLIENT_SOCKETS(cs) return true;
+	for (NetworkClientSocket *cs : NetworkClientSocket::Iterate()) return true;
```
Comment (glx22) [inline src/network/network.cpp]: I should probably rewrite this to get rid of unused variable warning

File: src/road_cmd.cpp [hunk 4]
```diff
@@ -182,8 +182,7 @@ RoadType AllocateRoadType(RoadTypeLabel label, RoadTramType rtt)
  */
 bool RoadVehiclesAreBuilt()
 {
-	const RoadVehicle *rv;
-	FOR_ALL_ROADVEHICLES(rv) return true;
+	for (const RoadVehicle *rv : RoadVehicle::Iterate()) return true;
```
Comment (glx22) [inline src/road_cmd.cpp]: I should probably rewrite this to get rid of unused variable warning

**Review by LordAro — APPROVED** [review]

**Pull request merged**